### ISSUE TITLE
[3.3.5] VMap data is sometimes not correct

### ISSUE THREAD

**Issue by jameyboor** [TrinityCore/TrinityCore#15161]
Rev : :link: https://github.com/TrinityCore/TrinityCore/commit/7d9843c380cf00f38ad0c5bc75fd387e50ee11da 

~~First thing: The Z position of collision enabled GO's isn't taken into account when updating the movement flags of creatures in `Creature::UpdateMovementFlags()`. It's "allowed" Z position is taken with an overloaded function of `Map::GetHeight()` that doesn't even use the dynamictree.
`float ground = GetMap()->GetHeight(GetPositionX(), GetPositionY(), GetPositionZMinusOffset());` uses `float Map::GetHeight(float x, float y, float z, bool checkVMap /*= true*/, float maxSearchDist /*= DEFAULT_HEIGHT_SEARCH*/) const` which doesn't take the dynamictree into account, it should be using `float Map::GetHeight(uint32 phasemask, float x, float y, float z, bool vmap/*=true*/, float maxSearchDist/*=DEFAULT_HEIGHT_SEARCH*/) const` so I've changed that.~~

Next thing, and my real issue: Some positions on certain GameObject's that are spawned in the air don't have correct FloorZ coords and they just use the FloorZ of the ground, which is GroundZ , this causes creatures that are aggro'd on GameObject's to fall through the GameObject because the core thinks the creature is flying. This doesn't happen to players because (I think) the client does have the correct height data of these GameObject's and the player's position is checked client side too, which allows them to stay on the GameObject.

An example of this problem is with the GameObject `Nexus Raid Platform` entry : `193070`
Me standing on the center of the platform using .gps : http://puu.sh/jcltV/a4e28abfed.jpg , this shows FloorZ at ~40 which is totally correct.
Me standing on a random spot somewhere else using .gps : http://puu.sh/jclrR/07d43fd663.jpg , this shows FloorZ at 0.5 which is not correct at all, it's most likely the FloorZ of the Ground beneath me.

Another example is with a creature spawned, the .gps information shown on the right is from the only creature on that platform, so is the `UpdateMovementFlags` call because I added a conditional breakpoint : http://puu.sh/jcniX/508acaad94.jpg

If anyone has some more information regarding this I'd love to hear it.

**Comment by jameyboor** [comment]
Note, this has nothing to do with phasemasks, this has to do with GO's in general.

**Comment by tkrokli** [comment]
I am guessing this can be related to  [[3.3.5/Ulduar] Mimiron and XT002 Entering the ground](https://github.com/TrinityCore/TrinityCore/issues/15032) #15032 ?

**Comment by Rochet2** [comment]
Maybe floor is paper thin on server side code making stuff that goes past the paper glitch..
Seemed to be the case with the objects when I tested. Simply jumping corrected the floorz and most erroring was when moving up slopes so when you most likely get past the "paper thin floor".

**Comment by Keader** [comment]
I do not know if it's related, but #15168 decreased the chances, but did not solve the problem

**Comment by jameyboor** [comment]
Indeed, it was a partial fix to such problems, the full fix is still WIP as said in the PR

**Comment by MrSmite** [comment]
> the .gps information shown on the right is from the only creature on that platform

Slight correction, the GPS of the spawn is actually yours. Notice the cell is [43,33] which is the same as your first screenshot (your second screenshot shows [42,33] after you moved).

**Comment by Aokromes** [comment]
This bug is still valid? a change on vmaps was made few weeks ago.

**Comment by Rochet2** [comment]
Tested on https://github.com/TrinityCore/TrinityCore/commit/3a27e2f7f0a5fa6fa2bf02bb2a5d975e7240ad44
and I spawned the gameobject entry 193070 and tried:
- spawning NPCs and using .gps on them.
- walking around and using .gps. especially tested the steep slopes on the gameobject. (stopped in the middle of them)
- running around at speed 10 and using .gps. especially tested the steep slopes on the gameobject. (stopped in the middle of them)
- spawning npcs and aggroing them and making them walk accross the platform and the steep slopes
- spawning npcs and setting a spawndist for them so they walk around the platform

All floorz were correct in the .gps command prints and no NPC fell through the platform.
So its possible this has been fixed.
